### PR TITLE
Print timing info on verbose

### PR DESCRIPTION
**Pull request by vladaionescu**
Fixes https://github.com/earthly/earthly/issues/529
Fixes https://github.com/earthly/earthly/issues/496

Prints elapsed time after completing each command and also a summary at the end for total time spent on each target.

Screenshot from 2020-11-20 16-47-01
Screenshot from 2020-11-20 16-47-14

CC @agbell @dchw

### CODE REVIEW

File: builder/solver_monitor.go [hunk 1]
```diff
@@ -264,6 +287,60 @@ func (sm *solverMonitor) printHeader(vm *vertexMonitor) {
...
+	dur := vertex.Completed.Sub(*vertex.Started)
+	if dur == 0 {
+		return
+	}
+	key := timingKey{
+		targetStr:      targetStr,
+		targetBrackets: targetBrackets,
+		salt:           salt,
+	}
+	sm.timingTable[key] = sm.timingTable[key] + dur
```
Comment (alexcb) [inline builder/solver_monitor.go:303]: not that it really matters, but is it not possible to do this instead?
```suggestion
	sm.timingTable[key] += dur
```

File: builder/solver_monitor.go [hunk 2]
```diff
@@ -264,6 +287,60 @@ func (sm *solverMonitor) printHeader(vm *vertexMonitor) {
...
+	if !sm.verbose {
+		return
+	}
+	sm.console.
+		WithMetadataMode(true).
+		Printf("Summary of timing information\n" +
+			"Note that the times do not include the expansion of commands like BUILD, FROM, COPY (artifact).")
+	var total time.Duration
+	durs := make([]time.Duration, 0, len(sm.timingTable))
+	durMap := make(map[time.Duration][]timingKey)
```
Comment (alexcb) [inline builder/solver_monitor.go:316]: I don't understand why the code is bucketing this by duration? is there a reason we can't just sort sm.timingTable directly (or make a copy of it, then sort it)? Then when it's printed to the console, one would only need to iterate over a single for loop rather than have two nested loops.

Comment (vladaionescu) [reply builder/solver_monitor.go:316]: Not really... maps in Go are not sorted. So we have to collect values in a slice and sort that slice and then reverse lookup. Don't know if you have something else in mind

File: builder/solver_monitor.go (hunk 1 above)
Comment (vladaionescu) [reply builder/solver_monitor.go]: :+1:

File: builder/solver_monitor.go (hunk 2 above)
Comment (vladaionescu) [reply builder/solver_monitor.go:315]: The double for loop is only for cases where multiple `timingKey`s have the exact same duration (very rare). It's not really inefficient.

Comment (alexcb) [reply builder/solver_monitor.go:315]: Ah, I didn't realise `sm.timingTable` was a map; I was wondering if we could do something like this instead? https://github.com/earthly/earthly/pull/533/files

I found the use of durMap hard to follow.

Comment (alexcb) [reply builder/solver_monitor.go:315]: full disclosure, I haven't tested that code; I'm also fine if you want to keep your code as-is.

Comment (vladaionescu) [reply builder/solver_monitor.go:315]: Yeah, much nicer. I would also move

```
type durationAndKey struct {
	dur time.Duration
	key timingKey
}
```

within the function itself since it's only used there.

**Comment by alexcb** [comment]
Sorry I missed a wee bit in that example branch of mine:
```
alex@mah:~/gh/earthly/earthly$ git diff
diff --git a/builder/solver_monitor.go b/builder/solver_monitor.go
index a04a27e..096c73c 100644
--- a/builder/solver_monitor.go
+++ b/builder/solver_monitor.go
@@ -328,6 +328,12 @@ func (sm *solverMonitor) PrintTiming() {
        sort.Slice(durs, func(i, j int) bool {
                return durs[i].dur > durs[j].dur
        })
+       for _, x := range durs {
+               sm.console.
+                       WithPrefixAndSalt(x.key.targetStr, x.key.salt).
+                       WithMetadataMode(true).
+                       Printf("(%s) %s\n", x.key.targetBrackets, x.dur)
+       }
        sm.console.
                WithMetadataMode(true).
                Printf("===============================================================\n")
```

**Comment by vladaionescu** [comment]
It should be good now

**Review by alexcb — APPROVED** [review]
